Test for external endpoint

microbe73 opened the pull request:
### New Features
One new test for the tools/external endpoint along with the mock data for it, also updated auth to allow for mock authentication

grugna commented:
@microbe73 are all the tests passing now? I remember you mentioning that some tests were not passing last time we talked. If any old test is not passing please update those.

@paulmurdoch19 could you review this PR and run the tests?

microbe73 commented:
@grugna When I made those changes in the other PR that was to try and get more of the tests to pass, right now when I run this there are 4 passing tests including the one I added, 7 failing and 1 error. I made the changes to the code so that it was down to 2 tests failing and 1 error, but it required changing the endpoints themselves and commenting out some of the things that were causing errors and I thought you did not want those changes, should I make those changes or leave it as is?

grugna commented:
@microbe73 I think you need to make changes to the data instead of the code. At least one of the test if I remember correctly had a problem with the test data and the assert. That one should be easy to fix. 
We know the code at time point in time is correct, so let's make sure to update the tests so that they are aligned to this code.
Could you list here the tests that are not passing?

microbe73 commented:
@grugna here is the current output when I run the tests
FAILED tests/endpoint.py::test_tools_counts_no_data - KeyError: 'cache'
FAILED tests/endpoint.py::test_tools_counts_correct_data - KeyError: 'cache'
FAILED tests/endpoint.py::test_tools_survival_no_data - assert {'message': '...data points.'} == {'message': '...data points.'}
FAILED tests/endpoint.py::test_tools_survival_incorrect_data - TypeError: 'NoneType' object is not iterable
FAILED tests/endpoint.py::test_tools_survival_correct_data - AssertionError: assert {'-1': {'coun...2162902123}]}} == {'message': '...data points.'}
FAILED tests/endpoint.py::test_tools_stats_no_data - AssertionError: assert [] == ['NODAL']
FAILED tests/endpoint.py::test_tools_stats_incorrect_data - AssertionError: assert [] == ['NODAL']
ERROR tests/endpoint.py::test_tools_counts_incorrect_data - KeyError: 'Short_DATA_PATH'

microbe73 commented:
The survival_no_data one is easily fixable the test expects the phrase "event free curve" in the error message but the actual error just says curve

microbe73 commented:
These are the current failing/error tests after updating the test messages and updating the endpoints to use the correct environment variable:
FAILED tests/endpoint.py::test_tools_counts_no_data - KeyError: 'cache'
FAILED tests/endpoint.py::test_tools_counts_correct_data - KeyError: 'cache'
FAILED tests/endpoint.py::test_tools_survival_correct_data - KeyError: 'age_at_disease_phase'
ERROR tests/endpoint.py::test_tools_counts_incorrect_data - KeyError: 'Short_DATA_PATH'

grugna commented:
For the cache one it is probably a test setup problem, since that key is initialized in the [api.py](https://github.com/microbe73/PcdcAnalysisTools/blob/46b59d590cc4458ccd5e1f2888e9fc4ebaa21a46/PcdcAnalysisTools/api.py#L96-L97) file.

microbe73 commented:
I moved the cache line in the api.py file and updated the endpoint to use the correct environment variable and now there are only 3 errors:
FAILED tests/endpoint.py::test_tools_counts_correct_data - AssertionError: assert [{'consortium...ect': 0}, ...] == [{'consortium...ect': 0}, ...]
FAILED tests/endpoint.py::test_tools_survival_correct_data - KeyError: 'age_at_disease_phase'
ERROR tests/endpoint.py::test_tools_counts_incorrect_data - KeyError: 'Short_DATA_PATH'

Moving the line felt somewhat arbitrary though and I don't think it fixes the actual issue of the test setting up incorrectly, so I don't know if it is a change I should commit or not

grugna commented:
Yes, I agree. Don't commit that change, instead try to understand why the test is not setup properly and how to set it up. 

As far as these three remaining errors:
- the assert should be easy to verify and fix with either the mock data and/or the assert value
- and the other two are probably 1 a config error and the other a mock data error.

microbe73 commented:
I actually believe that for the cache config variable, it should be copied again. Currently, it is only initialized in app_init in Pcdc/api.py, but this function is not called for testing, as I can see because there is already the code 
if app.mock_data == 'True':
    app_register_blueprints(app)
and app_register_blueprints is called in app_init as well, so I assume app_init is not meant to be called if we are using mock data, and so I should just copy the app.config["cache"] = {} from app_init into this conditional as well. 

There is also strange behavior with the test_tools_counts_correct_data. When I run all of the tests, it gives that assertion error. However, when I run the test individually, it gives the error 
FAILED tests/endpoint.py::test_tools_counts_correct_data - KeyError: 'molecular_abnormality_result'

This error makes sense as molecular_abnormality_result is not a key that appears anywhere in the json used for testing. The code that uses the molecular_abnormality_result is used to increment the molecular_analysis_count by 1, and in the json there is one subject that has a molecular analysis, and the assert statement also expects there to be only 1, so I assume that this is the one that is supposed to be counted, should I just add a key value pair with random information for this molecular_abnormality_result? 

For the test_tools_counts_incorrect_data, I am not sure what the incorrect data is supposed to be. The assert statement says
    assert [{"consortium":"total","molecular_analysis":1,"study":0,"subject":3},
            {'consortium': 'None', 'molecular_analysis': 0, 'study': 0, 'subject': 0},
            {"consortium":"missing","molecular_analysis":0,"study":0,"subject":0}] == response.json
, so it appears that there is supposed to be some sort of data here and not no data, since the test_tools_counts_no_data test does not have this assert statement, but then I don't know which data that is supposed to be. Short_DATA_PATH is not a key that was ever set by the environment variables; there is SHORT_DATA_PATH but this is used for the correct data test.

For the test_tools_survival_correct_data, there is an age_at_disease_phase key in the data, but it is a sub-key in the molecular_analysis dictionary. The code in the survival/__init__.py appears to be accessing data["timings"]["age_at_disease_phase"] key as the variables 
AGE_AT_DISEASE_PHASE = "timings.age_at_disease_phase"
DISEASE_PHASE = "timings.disease_phase"
are defined, split into
    node, age_at_disease_phase = AGE_AT_DISEASE_PHASE.split('.')
    node, disease_phase = DISEASE_PHASE.split('.')
, and then used as
        dict_tmp = each.get(node)
        if dict_tmp:
            for n in dict_tmp:
                if n.get(disease_phase) == "Initial Diagnosis" and n.get(age_at_disease_phase) and n.get(age_at_disease_phase) >= 0:
is run. 

However, despite the fact that "timings" is not a key that appears at all in the data, the error is not with n.get(disease_phase) even though this should be an empty dictionary from my understanding, which is quite strange. The stack trace for the error goes into the Pandas library, so it may be possible that the error comes from this return statement

    return (
        pd.DataFrame.from_records(guppy_data)
        .assign(
            omitted=lambda x:
                ((x[status_var].isna()) | (x[status_var] == 'Unknown') |
                 (x[time_var].isna()) | (x[time_var] < 0) |
                 (x[age_at_disease_phase].isna()) | (x[age_at_disease_phase] < 0)),
            status=lambda x:
                np.where(x["omitted"], None, x[status_var] == status_str),
            time=lambda x:
                np.where(x["omitted"], None, (x[time_var] - x[age_at_disease_phase]) / 365.25)      # TODO you can add an omit if   x[age_at_disease_phase]) > x[time_var]
        )
        .filter(items=["omitted", "status", "time"])
    )
, but this is quite confusing for me to read and I'm not entirely sure what it does.

Sorry that this message got quite long, I was writing out everything I knew in regards to these last few tests that fail when I run them.

grugna commented:
molecular_abnormality_result values are:
Present
Absent
Not Done
Not Reported
Unknown

add that as a variable inside the mock data file

grugna commented:
@microbe73 try using the following data for the survival test, and double check t doesn't break other things:

```
{
  "data": {
    "subject": [
      {
        "censor_status": "Subject is censored (i.e. has had no event(s))",
        "age_at_censor_status": 39,
        "survival_characteristics": [
          {
            "lkss": null,
            "age_at_lkss": 91
          }
        ],
        "timings": null
      },
      {
        "censor_status": "Subject is censored (i.e. has had no event(s))",
        "age_at_censor_status": 3,
        "survival_characteristics": [
          {
            "lkss": "Alive",
            "age_at_lkss": 85
          }
        ],
        "timings": null
      },
      {
        "censor_status": "Unknown",
        "age_at_censor_status": 18,
        "survival_characteristics": [
          {
            "lkss": "Dead",
            "age_at_lkss": 10
          }
        ],
        "timings": [
          {
            "age_at_disease_phase": 36,
            "disease_phase": "Relapse"
          }
        ]
      },
      {
        "censor_status": "Subject has had one or more events",
        "age_at_censor_status": 88,
        "survival_characteristics": [
          {
            "lkss": "Unknown",
            "age_at_lkss": 41
          }
        ],
        "timings": null
      },
      {
        "censor_status": "Subject is censored (i.e. has had no event(s))",
        "age_at_censor_status": 9,
        "survival_characteristics": [
          {
            "lkss": null,
            "age_at_lkss": 95
          }
        ],
        "timings": [
          {
            "age_at_disease_phase": 74,
            "disease_phase": "Relapse"
          },
          {
            "age_at_disease_phase": 90,
            "disease_phase": null
          },
          {
            "age_at_disease_phase": 96,
            "disease_phase": "Progression"
          }
        ]
      },
      {
        "censor_status": "Unknown",
        "age_at_censor_status": 70,
        "survival_characteristics": [
          {
            "lkss": null,
            "age_at_lkss": 30
          }
        ],
        "timings": [
          {
            "age_at_disease_phase": 33,
            "disease_phase": "Refractory"
          }
        ]
      },
      {
        "censor_status": null,
        "age_at_censor_status": 42,
        "survival_characteristics": null,
        "timings": [
          {
            "age_at_disease_phase": 90,
            "disease_phase": "Refractory"
          },
          {
            "age_at_disease_phase": 80,
            "disease_phase": "Initial Diagnosis"
          }
        ]
      },
      {
        "censor_status": "Subject is censored (i.e. has had no event(s))",
        "age_at_censor_status": 30,
        "survival_characteristics": null,
        "timings": null
      },
      {
        "censor_status": "Subject is censored (i.e. has had no event(s))",
        "age_at_censor_status": 37,
        "survival_characteristics": null,
        "timings": null
      },
      {
        "censor_status": "Subject is censored (i.e. has had no event(s))",
        "age_at_censor_status": 94,
        "survival_characteristics": [
          {
            "lkss": null,
            "age_at_lkss": 52
          }
        ],
        "timings": null
      }
    ]
  }
}
```

grugna commented on PcdcAnalysisTools/blueprint/routes/views/counts/__init__.py:
```diff
@@ -47,7 +47,8 @@ def get_result():
     # Cache is invalidated if the service is restarted or if the reset_count_cache endpoint is used
     capp.logger.info("CACHING COUNTS VALUE - " + json.dumps(counts))
-    capp.config["cache"]["counts"] = counts
+    if(capp.mock_data != 'True'):
```
If the image here is initialized correctly, I don't think you need to add this if statement to skip this.
<img width="536" alt="image" src="https://github.com/user-attachments/assets/33d333a5-7d91-471e-a2e2-0111839e6b9e">

microbe73 replied:
In lines 27-30 of the get_result function in .../views/counts/__init__.py:
`    if cache:
        if "counts" in cache and cache["counts"] is not None:
            capp.logger.info("RETRIEVING CACHE COUNTS VALUE - " + json.dumps(cache["counts"]))
            return flask.jsonify(cache["counts"])
`
If I get rid of the if statement to skip the cache, then two of the tests return errors because instead of accessing their own data they access the same data as the first counts test, and then there is an assertion error since each test is supposed to be accessing different sets of data.

grugna replied:
Then you want to reset the cache before each tests instead. The if statement in the application code is a little weird for this. 
You can do something like this - https://flask.palletsprojects.com/en/1.1.x/testing/#accessing-and-modifying-sessions
Or something like this: https://stackoverflow.com/questions/73865462/reset-global-variables-between-two-tests-on-pytest#answer-73872448

I am sure there are other solutions, these are just the first two I found with a quick google search.

microbe73 replied:
I added a test fixture to reset the cache and removed the conditional from the application code, all the tests are now passing

paulmurdoch19 commented on tests/endpoint.py:
```diff
@@ -130,6 +138,11 @@ def stats_correct_data(set_data):
 
+@pytest.fixture()
+def external_correct_data(set_data):
+    set_data('SHORT_DATA_EXTERNAL_PATH')
```
@microbe73 @grugna The readme should be updated explaining the new variables.

grugna replied:
You are right, thank you @paulmurdoch19.
@microbe73  can you push those changes.
